File: sketches/sketches.model/src/main/java/edu/kit/kastel/lissa/sketches/model/impl/elements/uml/UMLThing.java
package edu.kit.kastel.lissa.sketches.model.impl.elements.uml;

import org.eclipse.collections.api.list.ImmutableList;

import edu.kit.kastel.lissa.sketches.model.elements.uml.IUMLThing;
import edu.kit.kastel.lissa.sketches.model.impl.Box;
import edu.kit.kastel.lissa.sketches.util.ListWrapper;

abstract class UMLThing extends Box implements IUMLThing {
	private static final long serialVersionUID = -8954737446536608236L;

	private ListWrapper<String> attributes;
	private ListWrapper<String> methods;

	protected UMLThing() {
		this.attributes = ListWrapper.empty();
		this.methods = ListWrapper.empty();
	}

	@Override
	public void addAttribute(String attribute) {
		this.attributes.add(attribute);
	}

	@Override
	public void delAttribute(String attribute) {
		this.attributes.remove(attribute);
	}

	@Override
	public ImmutableList<String> getAttributes() {
		return this.attributes.toImmutable();
	}

	@Override
	public void addMethod(String method) {
		this.methods.add(method);
	}

	@Override
	public void delMethod(String method) {
		this.methods.remove(method);
	}

	@Override
	public ImmutableList<String> getMethods() {
		return this.methods.toImmutable();
	}
}

File: sketches/model/src/main/kotlin/edu/kit/kastel/lissa/sketches/model/impl/elements/uml/UMLThing.kt
package edu.kit.kastel.lissa.sketches.model.impl.elements.uml

import edu.kit.kastel.lissa.sketches.model.elements.uml.IUMLThing
import edu.kit.kastel.lissa.sketches.model.impl.Box

abstract class UMLThing protected constructor() : Box(), IUMLThing {
    private val attributes: MutableList<String> = mutableListOf()
    private val methods: MutableList<String> = mutableListOf()
    private var umlId: String? = null

    override fun addAttribute(attribute: String) {
        attributes.add(attribute)
    }

    override fun delAttribute(attribute: String) {
        attributes.remove(attribute)
    }

    override fun attributes(): List<String> {
        return attributes
    }

    override fun addMethod(method: String) {
        methods.add(method)
    }

    override fun delMethod(method: String) {
        methods.remove(method)
    }

    override fun methods(): List<String> {
        return methods
    }

    override fun setUMLId(umlId: String) {
        this.umlId = umlId
    }

    override fun umlId(): String {
        return umlId!!
    }
}
